File: sorting_python/sort.py
from os import path
from pickle import TRUE
# /**
#  * @brief 
#  * 
#  * This should be the base class for all sorting algorithms you implement. 
#  * Any sorting technique(like MergeSort, QuickSort etc) should be inherited from this base class.
#  * Use templates to support different data types(int, float, char).
#  */
class Sort:

       
        __data = None
        __data_len = None
    
        # /**
        #  * @brief 
        #  * Read data to sort from a file, whose name is passed as argument to the function.
        #  * Implement this function's body in src/sort.cpp file.
        #  * @return true, if file is opened successfully
        #  * @return false, if file any error while opening file
        #  */
        def read_data_from_file(self,filename):
            if not path.exists(filename):
                return False
            self.__data=[]
            fin = open(filename, 'r')       
            temp = fin.readlines()
            fin.close()
            for i in temp:
                if '.' in i:
                    self.__data.append(float(i.strip()))
                else:
                    self.__data.append(int(i.strip()))
                #self.__data.append((i.strip()))
            self.__data_len=len(self.__data)
            print(self.__data)
            print(self.__data_len)

            return True
            
        # /**
        # * @brief Construct a new sort data object
        # * Function to sort the data read from file into arr. 
        # * Implement the actual sorting operation in the derived class.
        # */
        def sort_data():
            pass
        # /**
        # * @brief 
        # * Write the sorted data to a file whose name is passed as argument.
        # * @param filename , output file name to save the sorted data into.
        # */
        def write_sorted_data_to_file(self,filename):
            fout = open(filename, 'w')
            print("In write_sorted_data_to_file")
            print(self.__data)
            for i in self.__data:
                fout.write(str(i))
                fout.write('\n') 

            fout.close()
        # Function to get the private members of the class
        def getData(self):
            return self.__data    

        # Function to modify the private members of the class
        def setData(self,array): 
            self.__data = array

File: sorting_cpp/src/sort.cpp
#include<iostream>
#include<fstream>
#include<string.h>
#include "../include/sort.hpp"
using namespace std;
template class Sort<int>;
template class Sort<std::string>;
template<class T>
bool Sort<T>::read_data_from_file(const char *filename){
    // Implement the code to read data from the file and store it in the data variable.
    int n=0,i=0;
    string line;
    ifstream fin;
    fin.open(filename);
    if (!fin) 
        return false;
    while(fin >>line)
        n++;
    data_len=n;
    fin.close();
    fin.open(filename);
    data=new T[n];
     while(fin >> line)
     {
        data[i]=stoi(line);
        i++;
     }
    fin.close();
    return true;


}
template<class T>
void Sort<T>::write_sorted_data_to_file(const char *filename){
    // Implement the code to write sorted data to the file.
    int i;
    ofstream fout;
    fout.open(filename);
    if(!fout)
        cout<<"error"<<endl;
    for (i=0; i<data_len; i++)
      fout << data[i] << endl;
   fout.close();
    

}
template<class T>
T* Sort<T>::getd()
{
         return data;
}
template<class T>
int Sort<T>::getl()
{
        return data_len;
}

template<class T>
void Sort<T>::setd(T* val,int len)
{
    this->data=val;
    this->data_len=len;
     
}

// template<class T>
// void Sort<T>::print()
// {
//     for(int i=0;i<data_len;i++)
//         cout<<"THIS:"<<data[i]<<endl;
     
// }


